index.d.ts

```diff
@@ -746,9 +746,7 @@ declare module 'mongoose' {
   interface Model<T, TQueryHelpers = {}, TMethods = {}, TVirtuals = {}> extends NodeJS.EventEmitter, AcceptsDiscriminator {
     new(doc?: AnyKeys<T> & AnyObject, fields?: any | null, options?: boolean | AnyObject): EnforceDocument<T, TMethods, TVirtuals>;
 
-    aggregate<R = any>(pipeline?: any[], options?: Record<string, unknown>): Aggregate<Array<R>>;
-    aggregate<R = any>(pipeline: any[], cb: Function): Promise<Array<R>>;
-    aggregate<R = any>(pipeline: any[], options: Record<string, unknown>, cb: Function): Promise<Array<R>>;
+    aggregate<R = any>(pipeline?: PipelineStage<T>[], options?: mongodb.AggregateOptions, callback?: Callback<R[]>): Aggregate<Array<R>>;
```

Comment:
This change is incorrect, because we need to support the `aggregate(pipeline, cb)` syntax. Please add back `aggregate<R = any>(pipeline: any[], cb: Function): Aggregate<Array<R>>;`

Thanks for fixing the issue with returning `Promise` though, that is an issue in our type definitions.

Reply:
Will revert back the two params signature 👍 .

About performance, how about removing the generic T from `PipelineStage<T>`, as it only helps for autocompletion of only a few stages like `$match` ?
Or do you think that the problem is not about the generic but the whole _one property object_ big union ?